### PR TITLE
enable zftp Jes functions and optimize some code

### PR DESCRIPTION
**Pull request by tiantn**
Signed-off-by: tian na <tiantn@cn.ibm.com>

## Proposed changes

Enabled zftp Jes functions in this PR. Also optimize some code, like conmon used getSession().

## Release Notes

Milestone:

Changelog:

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [ x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [ ] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [ ] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [ ] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by tiantn** [comment]
The following tests was manually done：
List Jobs with prefix * owner *
List job by jobid
List spool files
View spool files content
Download spool files
Submit job
Delete job

File: packages/zowe-explorer-ftp-extension/src/abstractFtpApi.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/*
```
Comment (std4lqi) [inline packages/zowe-explorer-ftp-extension/src/abstractFtpApi.ts:1]: A better file name like `ZoweExplorerAbstractFtpApi.ts`.

File: packages/zowe-explorer-ftp-extension/src/abstractFtpApi.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ *                                                                                 *
+ * Copyright Contributors to the Zowe Project.                                     *
+ *                                                                                 *
+ */
+
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+import * as imperative from "@zowe/imperative";
+import { ZoweExplorerApi } from "@zowe/zowe-explorer-api";
+
+export class AbstractFtpApi implements ZoweExplorerApi.ICommon {
```
Comment (std4lqi) [inline packages/zowe-explorer-ftp-extension/src/abstractFtpApi.ts:16]: Add `abstract` before `class` to indicate it's an abstract class.

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+            id: 0,
+            ddname: "",
+            "records-url": "",
+            lrecl: 0,
+            subsystem: "",
+            stepname: "",
+            procstep: "",
+        };
+    }
+    private checkedProfile(): imperative.IProfileLoaded {
```
Comment (std4lqi) [inline packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts:229]: I think `checkedProfile()` and `ftpClient()` can be moved into `AbstractFtpApi` as protected methods as well.

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    public async submitJob(jobDataSet: string): Promise<zowe.IJob> {
+        const result = this.getIJobResponse();
+        const connection = await this.ftpClient(this.checkedProfile());
+        if (connection) {
+            const transferOptions = {
+                transferType: TRANSFER_TYPE_ASCII,
+            };
+            const content = await DataSetUtils.downloadDataSet(connection, jobDataSet, transferOptions);
+            const jcl = content.toString();
+            const response: string = await JobUtils.submitJob(connection, jcl);
```
Comment (std4lqi) [inline packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts:178]: `jobId` is more informative than `response`.

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                const destinationFile = DownloadJobs.getSpoolDownloadFile(
+                    mockJobFile,
+                    parms.omitJobidDirectory,
+                    parms.outDir
+                );
+                imperative.IO.createDirsSyncFromFilePath(destinationFile);
+                imperative.IO.writeFile(destinationFile, spoolFileToDownload.contents);
+            }
+        }
+        return;
```
Comment (std4lqi) [inline packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts:140]: Unnecessary return I think.

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    }
+    public async downloadSpoolContent(parms: zowe.IDownloadAllSpoolContentParms): Promise<void> {
+        const connection = await this.ftpClient(this.checkedProfile());
+        if (connection) {
+            const destination = parms.outDir == null ? "./output/" : parms.outDir;
+            const jobDetails = await JobUtils.findJobByID(connection, parms.jobid);
+            if (jobDetails.spoolFiles == null || jobDetails.spoolFiles.length === 0) {
+                throw new Error("No spool files were available.");
+            }
+            const fullSpoolFiles = await JobUtils.getSpoolFiles(connection, jobDetails.jobid);
```
Comment (std4lqi) [inline packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts:108]: As we discussed, it's duplicate code with zftp. We may add new job API in the next zftp to cover spool file downloading.

**Review by std4lqi — COMMENTED** [review]
It looks very good. Just some minor comments.

File: packages/zowe-explorer-ftp-extension/src/abstractFtpApi.ts (hunk 1 above)
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/abstractFtpApi.ts]: Thanks Qi! I updated the name to ZoweExplorerAbstractFtpApi.ts.

File: packages/zowe-explorer-ftp-extension/src/abstractFtpApi.ts (hunk 2 above)
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/abstractFtpApi.ts]: done.

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts (hunk 3 above)
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts]: Thanks @Qi! I moved them to ZoweExplorerAbstractFtpApi.ts

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts (hunk 4 above)
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts]: done.

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts (hunk 5 above)
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts]: I removed the return. Thanks!

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts (hunk 6 above)
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts:109]: Yes, I add a comment in this part. Thank you!

**Review by lauren-li — COMMENTED** [review]
Overall, this PR seems to work very well! All unit and expected integration tests are passing (although this PR shouldn't really affect those since they were created for the core Zowe Explorer package). I was able to successfully perform the JES functionalities listed in comment https://github.com/zowe/vscode-extension-for-zowe/pull/1252#issuecomment-809194104 with my zFTP profile. The error handling for unsupported functions also seems covered.

I left a couple minor suggestions for making the error messages more specfic to the FTP extension.  

One additional request I have is to consider addressing my comment from the FTP for MVS PR: https://github.com/zowe/vscode-extension-for-zowe/pull/1219/files#r592555908. It doesn't have to be in this PR, but I think it would improve the user experience with the FTP extension.

Thank you @tiantn for your hard work on this PR!

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            if (response) {
+                // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+                return response.toString();
+            }
+        }
+        return "";
+    }
+
+    public getJclForJob(job: zowe.IJob): Promise<string> {
+        throw new Error("Get jcl is not supported.");
```
[the remark this answers is not in the document]
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts]: Thank you  @lauren-li! I updated the error message.

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    public submitJcl(jcl: string, internalReaderRecfm?: string, internalReaderLrecl?: string): Promise<zowe.IJob> {
+        throw new Error("Submit jcl is not supported.");
```
[the remark this answers is not in the document]
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpJesApi.ts]: Thank you  @lauren-li! I updated the error message.

**Review by lauren-li — APPROVED** [review]
Thanks @tiantn for addressing my comments. This PR looks good to me now!

**Comment by jellypuno** [comment]
I wanted to test this PR but unfortunately I am unable to connect to FTP 😢 But I did notice 2 things:
1. `Zowe Explorer was modified for FTP support.` is displayed twice
2. I cannot add my ftp profile in the Jobs Tree. Is this just me? I re-cloned the repo, fetch the branch then run yarn and yarn run build but I still cannot add the profile.

**Comment by tiantn** [comment]
> I wanted to test this PR but unfortunately I am unable to connect to FTP 😢 But I did notice 2 things:
> 
>     1. `Zowe Explorer was modified for FTP support.` is displayed twice
> 
>     2. I cannot add my ftp profile in the Jobs Tree. Is this just me? I re-cloned the repo, fetch the branch then run yarn and yarn run build but I still cannot add the profile.

Thank you @jellypuno for trying this. I am not sure if it related to the version of VS Code. My VS Code version is February 2021 (version 1.54). And I retry this PR with the following steps and it works well for me.
1. git clone --single-branch --branch zftp-jes git@github.com:tiantn/vscode-extension-for-zowe.git
2. cd vscode-extension-for-zowe/packages
3. yarn
4. yarn packages
5. install both vscode-extension-for-zowe-1.13.1.vsix and zowe-explorer-ftp-extension-1.13.1.vsix from VSIX.
6. VS Code prompts you to restart. Follow it to restart.
7. After restart loaded, the message `Zowe Explorer was modified for FTP support` displayed once.
8. Create zftp profile.
9. Add frofile from jobs.

Could you help to retry it? Thank you and have a good luck!😊

**Comment by jellypuno** [comment]
@tiantn That worked. Thank you! 

The feature works great for me. One error that I encountered is when adding the job spool in favorites. To re-create:
1. Submit a job
2. Click the Job ID from the message
3. Click the Star Icon
4. Error Message appears:
image

But once I refresh the tree, I can add the job spool to favorites. This only happens during the scenario above.

**Comment by tiantn** [comment]
Thank you @jellypuno! I receated the problem you mentioned. It only appears the first time you use it. @std4lqi and I have debuged it and find in zowe explorer code, when getting node's profile node in favorites the node is null. But so far we don't clear why the node is null. Could you help to debug it? Thank you very much!

**Comment by lauren-li** [comment]
@tiantn @jellypuno I was able to recreate the error described above by following Jelly's steps. I tried to debug this a bit, but unfortunately I did not get much further than confirming @tiantn and @std4lqi 's findings.

However, I also was able to recreate the issue on our Zowe Explorer 1.13.1 release using a z/OSMF profile. As this seems to be a pre-existing issue and not specific to the FTP implementation, I believe it is reasonable to address this in a separate PR. I have filed issue https://github.com/zowe/vscode-extension-for-zowe/issues/1265 to target this bug. If I have any additional findings, I will post them there. Thanks for catching this!

**Comment by jellypuno** [comment]
@lauren-li Thank you! I can approve this and merge this but I will wait for your go signal. so far everything is good for me.

**Comment by lauren-li** [comment]
@jellypuno I think this PR is good to merge in. Go for it! 😄

**Comment by tiantn** [comment]
Thank you @lauren-li and @jellypuno helping to debug the problem and open the issue.

**Comment by jellypuno** [comment]
@lauren-li @zFernand0 @crawr @JillieBeanSim All checks are successful. I can merge this. Requesting for your approvals please.

**Comment by zFernand0** [comment]
> All checks are successful. I can merge this. Requesting for your approvals please.

Yarn update looks ok to me, haven't tested the full PR yet though.

**Comment by jellypuno** [comment]
@lauren-li @tiantn @zFernand0 do you know why the other github checks are not running? Theia tests and ZE tests are not triggered. Only the FTP ones

**Comment by zFernand0** [comment]
> do you know why the other github checks are not running? Theia tests and ZE tests are not triggered. Only the FTP ones

The ZE and Theia will only run if the changes are inside the `packages/zowe-explorer` folder.
If we want those to run on every PR, we just need to update the workflow.

**Comment by jellypuno** [comment]
> > do you know why the other github checks are not running? Theia tests and ZE tests are not triggered. Only the FTP ones
> 
> The ZE and Theia will only run if the changes are inside the `packages/zowe-explorer` folder.
> If we want those to run on every PR, we just need to update the workflow.

Understood. thank you. let me add that as a topic on standup

**Review by jellypuno — APPROVED** [review]

**Pull request merged**